source/includes/steps-fle-convert-to-a-remote-master-key-kmip.yaml

```diff
@@ -0,0 +1,132 @@
+           method.
+
+---
+title: Create a New Data Encryption Key
+ref: create-a-new-data-key
+content: |
+  To encrypt your data, you need a data encryption key generated from your
+  KMS-hosted **master key**. The following diagram shows the requests you need
+  to make from the client application to create and store a new **data
+  encryption key**:
+
+  .. image:: /figures/CSFLE_Data_Key_KMS.png
+     :alt: Diagram that describes creating a data encryption key when using a KMS provider
+
+  1. First, specify the following information to access the master key:
+
+     .. list-table::
+        :header-rows: 1
+        :stub-columns: 1
+        :widths: 30 15 45
+
+        * - Field
+          - Required
+          - Description
+
+        * - keyId
+          - No
+          - The ``keyId`` field of a 96 byte
+            `Secret Data managed object <http://docs.oasis-open.org/kmip/spec/v1.4/os/kmip-spec-v1.4-os.html#_Toc490660780>`__
+            stored in your KMIP KMS.
+
+            .. note:: Create a New Master Key
+
+               If you do not specify the ``keyId`` field in the ``masterKey`` document
+               you send to your KMIP KMS, the driver creates a new
+               96 Byte Secret Data managed object in your KMS to act as your
+               master key. 
+
+        * - endpoint
+          - Yes
+          - The URI of your KMIP KMS.
+
+  2. Once you have the required information, update and run the following code
+     to generate the new data encryption key:
+     
+  .. include:: /includes/substitute-placeholders.rst
+
+  .. tabs-drivers::
+
+     .. tab::
+        :tabid: java-sync
+
+        .. code-block:: java
+
+           Map<String, Map<String, Object>> kmsProviderProperties = new HashMap<>();
+           Map<String, Object> providerDetails = new HashMap<>();
+           providerDetails.put("endpoint", "<KMIP KMS URI>");
+           kmsProviderProperties.put(kmsProvider,  providerDetails);
+           String keyVaultCollection = "<MongoDB namespace where you store your keys>"
+
+           ClientEncryption clientEncryption = ClientEncryptions.create(ClientEncryptionSettings.builder()
+               .keyVaultMongoClientSettings(MongoClientSettings.builder()
+                   .applyConnectionString(new ConnectionString("mongodb://localhost:27017"))
+                   .build())
+               .keyVaultNamespace(keyVaultNamespace)
+               .kmsProviders(kmsProviders)
+               .build());
+
+           DataKeyOptions dataKeyOptions = new DataKeyOptions().masterKey(
+               new BsonDocument()
+                   .append("keyId", "<your KeyId>"));
+
+           BsonBinary dataKeyId = clientEncryption.createDataKey("kmip", dataKeyOptions);
+           String base64DataKeyId = Base64.getEncoder().encodeToString(dataKeyId.getData());
+
+           System.out.println("DataKeyId [base64]: " + base64DataKeyId);
```

Comment:
What do you do with the Base64-encoded data key?  I don't think that the format that is actually included in the JSON schema.  Looking around I see two different formats.  In the [documentation](https://docs.mongodb.com/manual/core/security-client-side-encryption/) it shows:

```
   "keyId" : [UUID("e114f7ad-ad7a-4a68-81a7-ebcb9ea0953a")],
```
while in driver specification tests I see:

```
          "keyId": [
            {
              "$binary": {
                "base64": "AAAAAAAAAAAAAAAAAAAAAA==",
                "subType": "04"
              }
            }
```

@kevinAlbs  what do you think?

Reply:
@kevinAlbs will definitely be able to speak to this much more definitively than me, but I believe lines 114-116 only serve to print output for readers. Similar code can be found in the [CSFLE sample app here](https://github.com/mongodb-university/csfle-guides/blob/ca655b99f6d295bd57991b8bf3af3823727bd033/java/src/main/java/com/mongodb/csfle/util/CSFLEHelpers.java#L200-L203). I believe the key is added on line 113.

Reply:
> @kevinAlbs will be able to speak to this much more definitively than me, but I believe lines 114-116 only serve to print output for readers.

I also interpret the purpose is to print output for readers.

Hex output seems preferable. Users can pass a hex string directly in `mongosh` to the `UUID` helper.

In the Java driver, it looks like construct a `BsonBinary` from a hex string or base64 string would require converting to a byte array.

Reply:
I see two alternative:

```
       Base64.getEncoder().encodeToString(dataKeyId.getData());
```

will output something like: `s79lEYaTRbyxOtlEkSP1yQ==`

```
        dataKeyId.asUuid().toString();
 ```

will output something like: `b3bf6511-8693-45bc-b13a-d9449123f5c9`

@kevinAlbs which is preferable?  Will the shell accept either?

Reply:
These both work in mongosh:
```
Enterprise rs0 [direct: primary] test> uuid1 = UUID("b3bf6511-8693-45bc-b13a-d9449123f5c9")
UUID("b3bf6511-8693-45bc-b13a-d9449123f5c9")
Enterprise rs0 [direct: primary] test> uuid2 = BinData(04, "s79lEYaTRbyxOtlEkSP1yQ==")
UUID("b3bf6511-8693-45bc-b13a-d9449123f5c9")
Enterprise rs0 [direct: primary] test> uuid1.toJSON() == uuid2.toJSON()
true
```

Reply:
So either seems fine.

Reply:
Let's go with the second one, so we don't have to import any other classes.

Reply:
Just made the edit. If this edit is approved I'll make a JIRA ticket to update how print Keys throughout our CSFLE content.